### PR TITLE
manubot cite: option to render references

### LINKED ISSUE #48: Manubot cite: ability to output formatted bibliography
Suggested by @slochower in https://github.com/greenelab/manubot/pull/42#issuecomment-410873408:

> A super handy addition would be an option to output a formatted citation using the built-in style.csl. One challenge is that we currently leverage Pandoc to tie together the JSON for the references and the CSL. I suppose we could make a wrapper that writes the JSON to a file, writes a temporary Markdown file with the citation ids, calls Pandoc to render the temporary file, and then prints the rendered Markdown to stdout. That may be too clumsy to implement, though.

Brilliant! I agree this would be super useful, for cases ranging prototyping to just wanted to generate a reference.

One option would be to use [citeproc-py](https://github.com/brechtm/citeproc-py). I am also guessing there is a pandoc / pandoc-citeproc solution that is more direct than having to mess with creating a source markdown file. However, that would create a non-python dependency, which may be difficult for many casual users.

### PR DESCRIPTION
**Pull request by dhimmel**
Work in progress. ~~Rebase on #49 once that is merged.~~

Closes #48

### CODE REVIEW

**Comment by dhimmel** [comment]
Getting a frustrating [build error on Appveyer](https://ci.appveyor.com/project/greenelab/manubot/build/1.0.94/job/7ck6s00h0y9n9hha):

```
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\appveyor\AppData\Local\Temp\1\pip-req-build-r0y0a80m\setup.py", line 75, in <module>
        'manubot': 'cite/*.lua',
      File "c:\python36\lib\site-packages\setuptools\__init__.py", line 129, in setup
        return distutils.core.setup(**attrs)
      File "c:\python36\lib\distutils\core.py", line 148, in setup
        dist.run_commands()
      File "c:\python36\lib\distutils\dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "c:\python36\lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "c:\python36\lib\site-packages\setuptools\command\egg_info.py", line 278, in run
        self.find_sources()
      File "c:\python36\lib\site-packages\setuptools\command\egg_info.py", line 293, in find_sources
        mm.run()
      File "c:\python36\lib\site-packages\setuptools\command\egg_info.py", line 524, in run
        self.add_defaults()
      File "c:\python36\lib\site-packages\setuptools\command\egg_info.py", line 560, in add_defaults
        sdist.add_defaults(self)
      File "c:\python36\lib\site-packages\setuptools\command\py36compat.py", line 34, in add_defaults
        self._add_defaults_python()
      File "c:\python36\lib\site-packages\setuptools\command\sdist.py", line 134, in _add_defaults_python
        for _, src_dir, _, filenames in build_py.data_files:
      File "c:\python36\lib\site-packages\setuptools\command\build_py.py", line 66, in __getattr__
        self.data_files = self._get_data_files()
      File "c:\python36\lib\site-packages\setuptools\command\build_py.py", line 83, in _get_data_files
        return list(map(self._get_pkg_data_files, self.packages or ()))
      File "c:\python36\lib\site-packages\setuptools\command\build_py.py", line 95, in _get_pkg_data_files
        for file in self.find_data_files(package, src_dir)
      File "c:\python36\lib\site-packages\setuptools\command\build_py.py", line 114, in find_data_files
        return self.exclude_data_files(package, src_dir, files)
      File "c:\python36\lib\site-packages\setuptools\command\build_py.py", line 198, in exclude_data_files
        files = list(files)
      File "c:\python36\lib\site-packages\setuptools\command\build_py.py", line 234, in <genexpr>
        for pattern in raw_patterns
      File "c:\python36\lib\distutils\util.py", line 125, in convert_path
        raise ValueError("path '%s' cannot be absolute" % pathname)
    ValueError: path '/' cannot be absolute
```

Will update with links

**Comment by dhimmel** [comment]
@agitter can you look into the Windows build error and check our AppVeyor config to see if you spot anything I'm missing.

Also @agitter & @slochower, this is ready for a first pass of review.

**Comment by agitter** [comment]
@dhimmel I didn't check the AppVeyor build yet, but I can confirm that I have the same errors when trying to pip install version edd12f1 on my Windows machine.

I can spend a few minutes looking at it now.  If I don't find anything, I'll look into it further in the next couple days and also do a review.

**Comment by agitter** [comment]
@dhimmel here is a `setup.py` workaround that let's me install manubot on my local Windows machine:

```
    # Specify additional patterns to match files
    package_data={
        'manubot': str(pathlib.Path('cite/*.lua')),
    },
```

On Windows, that produces the string `'cite\\*.lua'`.  Does this work in Linux and AppVeyor too?

**Comment by slochower** [comment]
@dhimmel will review later this afternoon.

**Comment by agitter** [comment]
@dhimmel I saw that as well in the latest build.  My local version of pandoc is ancient so I wasn't able to immediately experiment with workarounds on my machine.

**Comment by dhimmel** [comment]
> My local version of pandoc is ancient so I wasn't able to immediately experiment with workarounds on my machine.

No worries. I should be able to take it from here.

**Comment by dhimmel** [comment]
Had to change the Travis build to use `sudo` so that we could install a current enough version of Pandoc to pass our `manubot cite --render` tests. This slows down the builds. One option is to not install pandoc on Travis, and just let these tests get executed on AppVeyor only. Thoughts?

**Comment by agitter** [comment]
The Pandoc tests seem sensitive to the versions and configuration, so I'm reassured by explicitly testing them on both Travis and AppVeyor.  However, you're the one doing active development who will suffer from slower builds, so I defer to you.

It would be great to selectively disable the pandoc tests via commit message, similar to the discussion at https://github.com/travis-ci/travis-ci/issues/8861

**Comment by dhimmel** [comment]
> The Pandoc tests seem sensitive to the versions and configuration, so I'm reassured by explicitly testing them on both Travis and AppVeyor

Let's keep Travis using sudo install of pandoc and if slowness becomes a problem, we will reevaluate.

**Comment by slochower** [comment]
Why do you need to `sudo` to install the latest version of Pandoc on Travis? It should be possible to install to a loca (and cached) directory without root permissions, a la `dpkg -i package.deb --force-not-root --root=$HOME`.

**Comment by dhimmel** [comment]
> It might be possible to install to a local (and cached) directory without root permissions

That would be great! Tried your suggesting in https://github.com/dhimmel/manubot/commit/a4feca207ac31ebe2d1490b17acebfa29de78f8f, but it [failed](https://travis-ci.org/greenelab/manubot/builds/417520539#L448):

```
dpkg: error: requested operation requires superuser privilege
```

File: manubot/cite/__init__.py [hunk 1]
```diff
@@ -145,11 +170,99 @@ def add_subparser_cite(subparsers):
...
+    if not args.render:
+        write_file = args.output.open('w') if args.output else sys.stdout
+        with write_file:
+            json.dump(csl_list, write_file, ensure_ascii=False, indent=2)
+            write_file.write('\n')
+        return
+
+    # use Pandoc to render citations
+    if not args.format and args.output:
+        vars(args)['format'] = extension_to_format.get(args.output.suffix)
```
[the remark this answers is not in the document]
Comment (agitter) [reply manubot/cite/__init__.py]: Nevermind, I see what happens.  If the suffix isn't in the dict, `format` will be `None` and will be set to a default in the two lines below.

File: README.md [hunk 2]
```diff
@@ -112,8 +114,16 @@ positional arguments:
 
 optional arguments:
   -h, --help            show this help message and exit
-  --file FILE           specify a file to write CSL output, otherwise default
-                        to stdout
+  --render              Whether to render CSL Data into a formatted reference
```
[the remark this answers is not in the document]
Comment (dhimmel) [reply README.md]: https://github.com/greenelab/manubot/pull/51/commits/867635d7a90442b2817bcca928249372709039a8

File: manubot/cite/__init__.py [hunk 3]
```diff
@@ -118,17 +119,41 @@ def citation_to_citeproc(citation, prune=True):
...
+    )
+    parser.add_argument(
+        '--format',
+        choices=list(extension_to_format.values()),
+        help="Format to use for output file. If not specified, attempt to infer this from filename extension. Otherwise, default to plain",
+    )
+    parser.add_argument(
+        '--output',
+        type=pathlib.Path,
+        help='specify a file to write output, otherwise default to stdout',
```
[the remark this answers is not in the document]
Comment (dhimmel) [reply manubot/cite/__init__.py]: This is a tough decision... whether to capitalize first letter and whether to use periods. While the general convention seems to be to not capitalize and not use periods, our help messages are more comprehensive than most. Therefore, I'm starting to lean towards Capitalize and use periods. @agitter preference?

[the remark this answers is not in the document]
Comment (agitter) [reply manubot/cite/__init__.py]: I don't have a preference either way.  It only stands out to me when the capitalization and punctuation changes.  Capitalization with periods would be okay with me.

**Comment by dhimmel** [comment]
Regarding failure in https://github.com/greenelab/manubot/pull/51/commits/77890d030ca90d0c1168095f87de4500c476a520, [we can set](https://stackoverflow.com/a/48093971/4651668) `PYTHONIOENCODING: "UTF-8"` in the appveyor config to prevent this.

File: manubot/cite/__init__.py [hunk 4]
```diff
@@ -249,20 +254,47 @@ def cli_cite(args):
...
+def _check_pandoc_version(info, metadata, format):
+    """
+    Given info from _get_pandoc_info, check that Pandoc's version is sufficient
+    to perform the citation rendering command specified by metadata and format.
+    Please add additional minimum version information to this function, as its
+    discovered.
+    """
+    issues = list()
+    if format == 'jats' and info['pandoc version'] < (2,):
+        issues.append('--jats requires pandoc ≥ v2.0.')
```
[the remark this answers is not in the document]
Comment (dhimmel) [reply manubot/cite/__init__.py]: Done in https://github.com/greenelab/manubot/pull/51/commits/54a7d2a994bd0ab072c29db7acbae882f6554412

File: manubot/cite/__init__.py [hunk 5]
```diff
@@ -138,33 +139,37 @@ def add_subparser_cite(subparsers):
...
-        help="Format to use for output file. If not specified, attempt to infer this from filename extension. Otherwise, default to plain",
+        help="When --render, format to use for output file. "
+             "If not specified, attempt to infer this from filename extension. "
+             "Otherwise, default to plain.",
     )
     parser.add_argument(
         '--output',
         type=pathlib.Path,
-        help='specify a file to write output, otherwise default to stdout',
+        help='Specify a file to write output, otherwise default to stdout.',
```
[the remark this answers is not in the document]
Comment (dhimmel) [reply manubot/cite/__init__.py]: Yes will do in next PR along with moving some code around

**Comment by dhimmel** [comment]
> You could also consider adding pandoc to the manubot-dev development conda environment in the readme.

Done.

> Would installing pandoc through conda be a good workaround for the dpkg install that requires sudo?

Good idea to explore in the future if sudo dpkg becomes a problem. One thing is that I find installed conda environments on Travis to be a bit slow.

**Comment by dhimmel** [comment]
> I'll refer to the explanation in greenelab/manubot-rootstock#81 (comment) as documentation for the preference for the conda-forge channel.

It's because `pandoc` is only in conda-forge.

**Comment by agitter** [comment]
I got pandoc from the default Anaconda channel
https://anaconda.org/anaconda/pandoc

**Pull request merged**